### PR TITLE
Progressive parsing with StreamParser

### PR DESCRIPTION
**Pull request by jhy**
A **StreamParser** provides a progressive parse of its input. As each Element is completed, it is emitted via a Stream or Iterator interface. Elements returned will be complete with all their children, and an (empty) next sibling, if applicable.

Elements (or their children) may be removed from the DOM during the parse, for e.g. to conserve memory, providing a mechanism to parse an input document that would otherwise be too large to fit into memory, yet still providing a DOM interface to the document and its elements.

Additionally, the parser provides a `selectFirst(String query)` / `selectNext(String query)`, which will run the parser until a hit is found, at which point the parse is suspended. It can be resumed via another `select() `call, or via the `stream()` or `iterator()` methods.

Once the input has been fully read, the input Reader will be closed. Or, if the whole document does not need to be read, call `stop()` and `close()`.

The `document()` method will return the Document being parsed into, which will be only partially complete until the input is fully consumed.

A StreamParser can be reused via a new `parse(Reader, String)`, but is not thread-safe for concurrent inputs. New parsers should be used in each thread.

If created via `Connection.Response#streamParser()`, or another Reader that is I/O backed, the iterator and
 stream consumers will throw an `java.io.UncheckedIOException` if the underlying Reader errors during read.

The StreamParser interface is currently in **beta** and may change in subsequent releases. Feedback on the feature and how you're using it is very welcome via the jsoup discussions .

### Process a file in chunks

Assuming we have a file with a bunch of `<book>` chunks each with many `<chapter>` elements, but loading it all into the DOM at once might run out of memory. Process the file in chunks by iterating on `selectNext(cssquery)`:

```java
static void streamChunks(Path path) throws IOException {
    try (StreamParser streamer = DataUtil.streamParser(
        path, StandardCharsets.UTF_8, "https://example.com", Parser.xmlParser())) {

        Element el;
        var seenChunks = 0;
        while ((el = streamer.selectNext("book")) != null) {
            // do something more useful! The element will have all its children elements
            Elements chapters = el.select("chapter");
            el.remove(); // remove this chunk once used to keep DOM light and not run out of memory
            seenChunks++;
        }

        Document doc = streamer.document(); // the completed doc, will just be a shell
        log("Title", doc.expectFirst("title"));
        log("Seen chunks", seenChunks);
    }
}
```

### Parse just the meta data of a website

Assume we are building a link preview tool. All the data we need is in the head section of a page, and so there's no need to fetch and parse the complete page. This example will fetch a given URL, parse only the `<head>` contents and use those, and then cleanly close the request:

```java
static void selectMeta(String url) throws IOException {
    try (StreamParser streamer = Jsoup.connect(url).execute().streamParser()) {
        Element head = streamer.selectFirst("head");
        if (head == null) return;

        log("Title", head.select("title").text());
        log("Description", head.select("meta[name=description]").attr("content"));
        log("Image", head.select("meta[name=twitter:image]").attr("content"));
    }
}
```

### CODE REVIEW

File: src/main/java/org/jsoup/helper/HttpConnection.java [hunk 1]
```diff
@@ -950,22 +953,47 @@
...
 
+        @Override public StreamParser streamParser() throws IOException {
+            InputStream stream = prepareParse();
+            String baseUri = url.toExternalForm();
+            DataUtil.CharsetDoc charsetDoc = DataUtil.detectCharset(stream, charset, baseUri, req.parser());
+            // note that there may be a document in CharsetDoc as a result of scanning meta-data -- but as requires a stream parse, it is not used here. todo - revisit.
+
+            // set up the stream parser and rig this connection up to the parsed doc:
+            StreamParser streamer = new StreamParser(req.parser());
+            BufferedReader reader = new BufferedReader(new InputStreamReader(stream, charsetDoc.charset));
```
[the remark this answers is not in the document]
Comment (jhy) [reply src/main/java/org/jsoup/helper/HttpConnection.java:987]: The reader is passed into the StreamParser, which is Closeable and will close the input stream.

File: src/main/java/org/jsoup/parser/TreeBuilder.java [hunk 2]
```diff
@@ -56,43 +60,58 @@
...
         return doc;
     }
 
+    List<Node> parseFragment(String inputFragment, @Nullable Element context, String baseUri, Parser parser) {
+        initialiseParse(new StringReader(inputFragment), baseUri, parser);
+        List<Node> nodes = doParseFragment(inputFragment, context, baseUri, parser);
+        return nodes;
+    }
+
+    abstract List<Node> doParseFragment(String inputFragment, @Nullable Element context, String baseUri, Parser parser);
```
[the remark this answers is not in the document]
Comment (jhy) [reply src/main/java/org/jsoup/parser/TreeBuilder.java]: Fixed

[the remark this answers is not in the document]
Comment (jhy) [reply src/main/java/org/jsoup/parser/TreeBuilder.java]: Fixed

**Comment by jhy** [comment]
All: if you're interested in this feature, it would be great if you could try it out before the next release by installing a snapshot version and bashing on it. Please comment here with any feedback (what works / doesn't work) and any suggestions on the API.

See details for building a snapshot on the [download](https://jsoup.org/download) page.

**Comment by 821938089** [comment]
Missing parseBodyFragment.

Is it possible to implement the removal of unmatched nodes during the select process?

**Comment by jhy** [comment]
@821938089 can you give a step-by-step example of what you mean by removing unmatched nodes? Or a sample implementation of `selectNext(Evaluator eval)` that would do it? It seems like it would be easy to trash the document.

**Comment by 821938089** [comment]
There seems to be no way to do immediate removal of mismatched nodes.
I have an imperfect way to do this: design a finite node cache set queue, put each parsed node into it, and remove the oldest node if it adds to the limit.
If a matching node is found remove it from the queue.

**Comment by jhy** [comment]
It's not clear to me what you are trying to get - can you give me a step by step of the input -> selector -> parsed DOM -> removal?

If we just removed every non-matching node after a select, that's going to get a lot of collateral damage.

**Comment by 821938089** [comment]
If you have a very large document and the things you're interested in are scattered throughout the document or at the end of the document, you're going to cause OOM when parsing because the entire document will be in memory.
Your implementation of select looks like it's streaming, but it's using more and more memory.
So I was wondering if there is a way to remove the mismatched nodes to prevent OOM.

**Comment by jhy** [comment]
I've added fragment parse options (and a completeFragment to pick up the relevant nodes) with 1f1f72d1e89821c630dcfc35e1a0a7f653cc877b

**Comment by partick33** [comment]
I want to gradually convert the textNode content through streaming, but the container is OOM. I found that the parsed element html is getting bigger and bigger. I use the iterator mode. Is there any good way to solve it?

**Comment by partick33** [comment]
I expect <a>123</a> => <a>123456</a>,  user remove() api, result is null

**Comment by partick33** [comment]
> @partick33 Remove elements that have been processed.删除已处理的元素。

I expect 123 => 123456, user remove() api, result is null

**Pull request merged**